power_grid_model_c/power_grid_model_cpp/include/power_grid_model_cpp/handle.hpp

```diff
@@ -0,0 +1,86 @@
+// SPDX-FileCopyrightText: Contributors to the Power Grid Model project <powergridmodel@lfenergy.org>
+//
+// SPDX-License-Identifier: MPL-2.0
+
+#pragma once
+#ifndef POWER_GRID_MODEL_CPP_HANDLE_HPP
+#define POWER_GRID_MODEL_CPP_HANDLE_HPP
+
+#include "power_grid_model_c/handle.h"
+
+#include "basics.hpp"
+
+namespace power_grid_model_cpp {
+class PowerGridError : public std::exception {
+  public:
+    PowerGridError(const std::string& message) : message_(message) {}
+    const char* what() const noexcept override { return message_.c_str(); }
+
+  private:
+    std::string message_;
+};
+
+class PowerGridRegularError : public PowerGridError {
+  public:
+    using PowerGridError::PowerGridError;
+    static constexpr Idx error_code() { return PGM_regular_error; }
+};
+
+class PowerGridSerializationError : public PowerGridError {
+  public:
+    using PowerGridError::PowerGridError;
+    static constexpr Idx error_code() { return PGM_serialization_error; }
+};
+
+class PowerGridBatchError : public PowerGridError {
+  public:
+    struct FailedScenario {
+        Idx scenario;
+        std::string error_message;
+    };
+
+    PowerGridBatchError(std::string const& message, std::vector<FailedScenario> failed_scenarios_c)
+        : PowerGridError{message}, failed_scenarios_{std::move(failed_scenarios_c)} {}
+    static constexpr Idx error_code() { return PGM_batch_error; }
+    std::vector<FailedScenario> const& failed_scenarios() const { return failed_scenarios_; }
+
+  private:
+    std::vector<FailedScenario> failed_scenarios_;
+};
+
+class Handle {
+  public:
+    Handle() : handle_{PGM_create_handle()} {}
+
+    RawHandle* get() const { return handle_.get(); }
+
+    void check_error() const {
+        Idx error_code = PGM_error_code(handle_.get());
+        std::string error_message = error_code == PGM_no_error ? "" : PGM_error_message(handle_.get());
+        switch (error_code) {
+        case 0:
+            return;
+        case PGM_regular_error:
+            throw PowerGridRegularError{error_message};
+        case PGM_batch_error: {
+            Idx const n_failed_scenarios = PGM_n_failed_scenarios(handle_.get());
+            std::vector<PowerGridBatchError::FailedScenario> failed_scenarios(n_failed_scenarios);
+            auto const failed_scenario_seqs = PGM_failed_scenarios(handle_.get());
+            auto const failed_scenario_messages = PGM_batch_errors(handle_.get());
+            for (Idx i = 0; i < n_failed_scenarios; ++i) {
+                failed_scenarios[i] =
+                    PowerGridBatchError::FailedScenario{failed_scenario_seqs[i], failed_scenario_messages[i]};
+            }
+            throw PowerGridBatchError{error_message, std::move(failed_scenarios)};
+        }
+        case PGM_serialization_error:
+            throw PowerGridSerializationError{error_message};
+        }
+    }
+
```

Comment:
I think this suggestion is a good approach to prevent forgetting to use `check_error()`. As in the other possible suggestion, I think it may add unnecessary complexity to just avoid manually calling `get()`, as intellisense is already very good at reminding you. However, I will leave this open to remind myself to double check at the end in case I end up finding it good to implement.

Reply:
On another note, it is probably a good idea to rename `call_with` member function to something more expresive? Maybe `call_with_check`?

Reply:
possible, but the "call with" was inspired by Pythonic `with` clauses (which are context managers with an optional setup, winddown and fallback called during exception-handling). That's exactly what this is, but now with error-code-like exception functionality, rather than logic-flow-type exceptions.

Reply:
If we go this approach. Please also call `clear_error` in the `call_with`. Otherwise the error stays here and it will throw again next time even there is no error.

Reply:
> possible, but the "call with" was inspired by Pythonic `with` clauses (which are context managers with an optional setup, winddown and fallback called during exception-handling). That's exactly what this is, but now with error-code-like exception functionality, rather than logic-flow-type exceptions.

We can leave the naming as is then.



> If we go this approach. Please also call `clear_error` in the `call_with`. Otherwise the error stays here and it will throw again next time even there is no error.

Done.

Reply:
I think this can be resolved, right?